### PR TITLE
IndexAutoConstructor

### PR DESCRIPTION
**Pull request by flexatone**

### CODE REVIEW

**Comment by flexatone** [comment]
@chaburkland : thought you might want to see this as it expands on some of your work with `IndexHierarchy`.

File: static_frame/test/unit/test_frame.py [hunk 1]
```diff
@@ -9666,11 +9665,34 @@ def test_frame_set_index_hierarchy_f(self) -> None:
...
+        dt64 = np.datetime64
+        self.assertEqual(f2.to_pairs(),
+                ((2, (((dt64('2210-12-26'), dt64('164167')), 'ztsv'), ((dt64('2224-04-06'), dt64('43127')), 'zUvW'), ((dt64('2202-08-20'), dt64('7699')), 'zkuW'))), (3, (((dt64('2210-12-26'), dt64('164167')), 'z2Oo'), ((dt64('2224-04-06'), dt64('43127')), 'z5l6'), ((dt64('2202-08-20'), dt64('7699')), 'zCE3')))))
+        self.assertEqual(f2.index.name, (0, 1))
+
+    def test_frame_set_index_hierarchy_h(self) -> None:
+        IACF = IndexAutoConstructorFactory
+        f1 = ff.parse('s(3,4)|v(dtD, dtY, str, str)')
+        f2 = f1.set_index_hierarchy((0, 1),
+                index_constructors=(IACF('a'), IACF('b')),
```
Comment (chaburkland) [inline static_frame/test/unit/test_frame.py:9688]: The names don't populate as expected for some cases.

```python
10:58:23  In [18] f1.set_index_hierarchy((0, 1), index_constructors=(sf.IndexAutoConstructorFactory("a"), sf.IndexAutoConstructorFactory("b"))).index.name
10:59:48 Out [18] ('a', 'b')

10:59:48  In [19] f1.set_index_hierarchy((0, 1), index_constructors=(sf.IndexAutoConstructorFactory, sf.IndexAutoConstructorFactory("b"))).index.name
10:59:54 Out [19] (0, 1)

10:59:54  In [20] f1.set_index_hierarchy((0, 1), index_constructors=(sf.IndexAutoConstructorFactory("a"), sf.IndexAutoConstructorFactory)).index.name
11:00:02 Out [20] (0, 1)

```

File: static_frame/core/index_hierarchy.py [hunk 2]
```diff
@@ -759,8 +785,7 @@ def gen_columns() -> tp.Iterator[np.ndarray]:
                 )
 
     # --------------------------------------------------------------------------
-
-    def _create_blocks_from_self(self: IH) -> TypeBlocks:
+    def _to_type_blocks(self: IH) -> TypeBlocks:
```
Comment (chaburkland) [inline static_frame/core/index_hierarchy.py:788]: This is a much better name!

File: static_frame/core/index_hierarchy.py [hunk 3]
```diff
@@ -261,36 +263,42 @@ def _build_index_constructors(cls: tp.Type[IH],
             *,
             index_constructors: IndexConstructors,
             depth: int,
-            ) -> tp.Iterable[IndexConstructor]:
+            ) -> tp.Iterator[IndexConstructor]:
         '''
         Returns an iterable of `depth` number of index constructors based on user-provided `index_constructors`
+
+        Args:
+            dtype_per_depth: Optionall provide a dtype per depth to be used with ``IndexAutoConstructorFactory``.
```
Comment (chaburkland) [inline static_frame/core/index_hierarchy.py:271]:
```suggestion
            dtype_per_depth: Optionally provide a dtype per depth to be used with ``IndexAutoConstructorFactory``.
```

File: static_frame/core/index_hierarchy.py [hunk 4]
```diff
@@ -713,12 +728,14 @@ def _from_type_blocks(cls: tp.Type[IH],
...
             own_blocks: bool = False,
+            name_priority: bool = True,
             ) -> IH:
         '''
         Construct an :obj:`IndexHierarchy` from a :obj:`TypeBlocks` instance.
 
         Args:
-            blocks: a TypeBlocks instance
+            blocks: a TypeBlocks
+            name_priority: if True, setting the name from the ``name`` arguemnt superseedes the names via index_constructors; if False, names via index_constructor will get precedence.
```
Comment (chaburkland) [inline static_frame/core/index_hierarchy.py:738]:
```suggestion
            name_priority: if True, setting the name from the ``name`` argument supersedes the names via index_constructors; if False, names via index_constructor will get precedence.
```

**Review by chaburkland — COMMENTED** [review]
I have some small concern about the verbosity of the new name. Can it be renamed to something like: `IndexAutoConstructor`? Or `IndexAutoFactory`? I might be missing something about what the different components of the name you've chosen mean.

Also, I'm not quite sure the usefulness of the ability to optionally provide a name to `IndexAutoConstructorFactory`. Wouldn't it make more sense to provide an additional argument to `set_index_hierarchy` that allows users to override the names? Are there any other use cases I'm not thinking of?

File: static_frame/test/unit/test_frame.py (hunk 1 above)
Comment (flexatone) [reply static_frame/test/unit/test_frame.py:9688]: That is a good catch!

**Comment by flexatone** [comment]
Many thanks, @chaburkland , for your feedback.

We already have `IndexAutoFactory`: that is used as a value for `index` arguments to specify an auto incremented integer index. `IndexAutoConstructorFactory` is different in that it is not a value for `index` arguments, but a value for `index_constructor` arguments. Having "constructor" in the name is thus important to distinguish these. I originally was using `IndexAutoConstructor` but that felt awkward as it is not a constructor but a factory of constructors...

Index constructors generally expose a `name` argument, so that is why this factory takes that name argument and passes it on to the final constructor.

File: static_frame/core/index_auto.py [hunk 5]
```diff
@@ -11,28 +12,88 @@
...
 from static_frame.core.util import IndexInitializer
 from static_frame.core.util import NameType
+from static_frame.core.util import NAME_DEFAULT
 
+class IndexConstructorFactory:
+    def __call__(self,
+            labels: tp.Iterator[tp.Hashable],
+            *,
+            name: NameType = NAME_DEFAULT,
+            default_constructor: tp.Type['IndexBase'],
```
[the remark this answers is not in the document]
Comment (flexatone) [reply static_frame/core/index_auto.py:22]: Thanks, there were many of these!

File: static_frame/core/index_auto.py [hunk 6]
```diff
@@ -11,28 +12,88 @@
 from static_frame.core.util import CallableOrMapping
 from static_frame.core.util import IndexInitializer
 from static_frame.core.util import NameType
+from static_frame.core.util import NAME_DEFAULT
 
+class IndexConstructorFactory:
```
[the remark this answers is not in the document]
Comment (flexatone) [reply static_frame/core/index_auto.py:17]: Good point; I will go with `IndexConstructorFactoryBase`; I think it is more base class than protocol anyway.

File: static_frame/core/index_auto.py [hunk 7]
```diff
@@ -11,28 +12,88 @@
...
+    def __call__(self,
+            labels: tp.Iterator[tp.Hashable],
+            *,
+            name: NameType = NAME_DEFAULT,
+            default_constructor: tp.Type['IndexBase'],
+            ) -> IndexBase:
+        '''Partial the passed constructor with the ``name``.
+        '''
+        name = self._name if name is NAME_DEFAULT else name
+        return partial(default_constructor, name=name)(labels)
```
[the remark this answers is not in the document]
Comment (flexatone) [reply static_frame/core/index_auto.py:47]: good catch... a left-over from previous iterations.

**Pull request merged**